File: src/main/java/cube8540/oauth/authentication/credentials/security/UriSecurityMetadataSource.java
package cube8540.oauth.authentication.credentials.security;

import lombok.Getter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.ConfigAttribute;
import org.springframework.security.web.FilterInvocation;
import org.springframework.security.web.util.matcher.RequestMatcher;
import org.springframework.stereotype.Component;

import javax.servlet.http.HttpServletRequest;
import java.util.Collection;
import java.util.Map;
import java.util.stream.Collectors;

@Getter
@Component
public class UriSecurityMetadataSource implements ReloadableFilterInvocationSecurityMetadataSource {

    private final SecurityMetadataLoadService service;

    private Map<RequestMatcher, Collection<ConfigAttribute>> metadata;

    @Autowired
    public UriSecurityMetadataSource(SecurityMetadataLoadService service) {
        this.service = service;
        this.metadata = service.loadSecurityMetadata();
    }

    @Override
    public Collection<ConfigAttribute> getAttributes(Object object) throws IllegalArgumentException {
        HttpServletRequest request = ((FilterInvocation) object).getRequest();
        return metadata.entrySet().stream().filter(entry -> entry.getKey().matches(request))
                .map(Map.Entry::getValue).flatMap(Collection::stream).collect(Collectors.toSet());
    }

    @Override
    public Collection<ConfigAttribute> getAllConfigAttributes() {
        return metadata.values().stream().flatMap(Collection::stream).collect(Collectors.toSet());
    }

    @Override
    public boolean supports(Class<?> clazz) {
        return FilterInvocation.class.isAssignableFrom(clazz);
    }

    @Override
    public void reload() {
        this.metadata = service.loadSecurityMetadata();
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/security/UriSecurityMetadataSource.kt
package cube8540.oauth.authentication.security

import org.springframework.security.access.ConfigAttribute
import org.springframework.security.web.FilterInvocation
import org.springframework.security.web.util.matcher.RequestMatcher
import org.springframework.stereotype.Component
import java.util.stream.Collectors

@Component
class UriSecurityMetadataSource(private val service: SecurityMetadataLoadService): ReloadableFilterInvocationSecurityMetadataSource {

    final var metadata: Map<RequestMatcher, Collection<ConfigAttribute>> = service.loadSecurityMetadata()
        private set

    override fun reload() {
        this.metadata = service.loadSecurityMetadata()
    }

    override fun getAttributes(`object`: Any?): MutableCollection<ConfigAttribute> {
        val request = (`object` as FilterInvocation).request

        return metadata.entries
            .filter { entry -> entry.key.matches(request) }
            .map { entry -> entry.value }
            .flatMap { entries -> entries.asIterable() }
            .toMutableSet()
    }

    override fun getAllConfigAttributes(): MutableCollection<ConfigAttribute> =
        metadata.values
            .flatMap { values -> values.asIterable() }
            .toMutableSet()

    override fun supports(clazz: Class<*>): Boolean = FilterInvocation::class.java.isAssignableFrom(clazz)
}